### PR TITLE
SAK-48287 - Support Time Zone Shifting for Due Dates

### PR DESCRIPTION
**Pull request by csev**

### CODE REVIEW

**Comment by csev** [comment]
Hey @ern this has date stuff so it could use a close look.

File: basiclti/tsugi-util/src/test/org/tsugi/basiclti/BasicLTIUtilTest.java [hunk 1]
```diff
@@ -557,4 +557,21 @@ public void testMergeCSV() {
 		retval = BasicLTIUtil.mergeCSV("1,2", "2,1,3,4", null);
 		assertEquals(retval, "1,2,3,4");
 	}
+
+	@Test
+	public void testShiftJVMDateToTimeZone() {
+		// tz=sun.util.calendar.ZoneInfo[id="America/New_York",offset=-18000000,...
+		TimeZone tz = TimeZone.getTimeZone("America/New_York");
+		assertEquals(tz.getRawOffset(), -18000000);
```
Comment (ottenhoff) [inline basiclti/tsugi-util/src/test/org/tsugi/basiclti/BasicLTIUtilTest.java:565]: what happens during DST?

File: basiclti/tsugi-util/src/java/org/tsugi/basiclti/BasicLTIUtil.java [hunk 2]
```diff
@@ -1336,6 +1336,24 @@ public static Date parseIMS8601(String timestamp) {
...
+	 */
+	public static Date shiftJVMDateToTimeZone(Date date, String timeZone) {
+		if ( date == null || timeZone == null ) return null;
+
+		// Get the JVM TimeZone
+		TimeZone tzJVM = TimeZone.getDefault();
+		TimeZone tzNew = TimeZone.getTimeZone(timeZone);
+		long dueTime = date.getTime();
+		// Shift into UTC and then back to the destination timezone
+		dueTime = dueTime - tzJVM.getRawOffset() + tzNew.getRawOffset();
```
Comment (ottenhoff) [inline basiclti/tsugi-util/src/java/org/tsugi/basiclti/BasicLTIUtil.java:1352]: yuck!

maybe this is low-stakes code (e.g., the exact time isn't super important). but if precision is important, this should really be done by the experts:

https://www.baeldung.com/java-date-to-localdate-and-localdatetime

Comment (csev) [reply basiclti/tsugi-util/src/java/org/tsugi/basiclti/BasicLTIUtil.java:1352]: Sam - In regards to the accuracy, these are due *dates* in Sakai - which are not even date time values let alone millisecond values.  Just date.  

I read the site you mention and none of the code in the site you reference talks about _**shifting**_ into a destination timezone different that the current server's time zone.  Which is the only problem that I need to solve here.

File: basiclti/tsugi-util/src/test/org/tsugi/basiclti/BasicLTIUtilTest.java (hunk 1 above)
Comment (csev) [reply basiclti/tsugi-util/src/test/org/tsugi/basiclti/BasicLTIUtilTest.java:565]: If things go wonky, then the unit test breaks and we know about it.  And fix the unit test.   If this code breaks it is good to know about that.  Part of what I put in these unit tests was my worry about things shifting around on us in future versions of the JVM, etc.  I worry too - so I want an obvious way to track that to call attention to when the rules change.

File: basiclti/tsugi-util/src/java/org/tsugi/basiclti/BasicLTIUtil.java (hunk 2 above)
Comment (ottenhoff) [reply basiclti/tsugi-util/src/java/org/tsugi/basiclti/BasicLTIUtil.java:1352]: ZonedDateTime ldtZoned = ldt.atZone(ZoneId.systemDefault());

ZonedDateTime utcZoned = ldtZoned.withZoneSameInstant(ZoneId.of("UTC"));

File: basiclti/tsugi-util/src/test/org/tsugi/basiclti/BasicLTIUtilTest.java (hunk 1 above)
Comment (ottenhoff) [reply basiclti/tsugi-util/src/test/org/tsugi/basiclti/BasicLTIUtilTest.java:565]: ok i'm fine with it.... just hurts to see you using all of the old bad practices ;)

**Comment by csev** [comment]
@ottenhoff - This is a super weird use case.  One SakaiPlus server in one time zone is sending dates from the JVM-local time zone that display in correctly in the SakaiPlus UI in the JVM local time zone, to other LMS servers (i.e. like Canvas in Pacific time).  When those servers receive the UTC they convert it into their local time zone and then display it in their local time zone.  So we need to adjust the UTC values we send so that when a date looks right in the SakaiServer in UTC - it *also* looks right in the Canvas in US/Eastern.  

It is both very weird and kind of cool.   SakaiPlus makes Sakai a "web service" end point and it needs to interact with tenants scattered around the world :)  And in the Plus Admin tool - you will see that every tenant can specify a different time zone.

This was never a problem when LTI tools let the calling LMS handle the due dates.  But SakaiPlus needs to *transfer* due dates - things like Cengage need to do this if they want to set due dates in their calling LMS's.

**Comment by ottenhoff** [comment]
sending dates to other systems in UTC makes perfect sense. 

java's old date handling was well-known for being pretty horrible. all the new java.time packages are pretty wonderful and handle every use case I've ever seen

File: basiclti/tsugi-util/src/java/org/tsugi/basiclti/BasicLTIUtil.java (hunk 2 above)
Comment (csev) [reply basiclti/tsugi-util/src/java/org/tsugi/basiclti/BasicLTIUtil.java:1352]: The ZonedDateTime really is obsessed with changing what is printed and shown.  I tried code like:

`ZonedDateTime ldtZoned = ZonedDateTime.ofInstant(date.toInstant(), ZoneId.systemDefault());
 ZonedDateTime shiftZoned = ldtZoned.withZoneSameLocal(ZoneId.of(timeZone));
 retval = java.util.Date.from(shiftZoned.toInstant());`

And also with `withZoneSameInstant()` and they did not actually shift after being converted back in a JVM Date and Converted to ISO8601 /l UTC.  

I tried to use the fancy code :)

**Pull request merged**